torch/_inductor/fx_passes/reinplace.py

```diff
@@ -566,16 +584,132 @@ def tensor_with_same_storage_already_reinplaced(arg):
                     replace_dict[copy_node] = copy_node.args[0]
                 node.target = inplaceable_op.inplace_op
         elif node.target == torch.ops.higher_order.auto_functionalized:
+            from torch._higher_order_ops.auto_functionalize import (
+                deserialize_views_meta,
+                get_mutable_args,
+            )
+
             _mutable_op = node.args[0]
-            from torch._higher_order_ops.auto_functionalize import get_mutable_arg_names
+            tensors_to_clone, arg_types = get_mutable_args(_mutable_op)
```

Comment:
Right now, we're formulating reinplacing in terms of "can we reinplace on the mutable args". Instead, we should probably formulate it as "can we replace on the bases". I think this would simplify the logic below to the point where we can use the reinplacing logic before this PR.

We can reinplace on a base if there are no reads from it later on

Reply:
So, concretely, I want "tensors_to_clone" to just be the bases and I want the reinplacing pass to decide which of the bases don't need to be cloned.

Reply:
I think it might work fine what you propose